nemo_aligner/models/nlp/gpt/megatron_gpt_rs_actor.py

```diff
@@ -0,0 +1,337 @@
+from nemo_aligner.models.alignable_interface import AlignableGenerativeInterface
+from nemo_aligner.utils.distributed import broadcast_2d_tensor_within_pp, from_parallel_logits_to_logprobs
+from nemo_aligner.utils.text_generation_utils import TrackLengthGPTModelTextGenerationStrategy
+from nemo_aligner.utils.train_utils import (
+    grad_reductions,
+    prepare_for_training_step,
+    set_eval,
+    set_sync_funcs,
+    set_train,
+)
+from nemo_aligner.utils.utils import (
+    adapter_control,
+    configure_batch_sizes,
+    cpu_weight_swap,
+    masked_mean,
+    offload_distributed_adam,
+)
+
+
+class MegatronGPTRSModel(NLPAdapterModelMixin, MegatronGPTModel, AlignableGenerativeInterface):
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer=trainer)
+        self.automatic_optimization = False
+
+        self.init_policy_state_dict = None
+        self.distributed_adam_offload_manager = None
+
+        # length parameters for generation
+        self._length_params = OmegaConf.to_container(self.cfg.rs.length_params, resolve=True)
+        # sampling parameters for generation
+        self._sampling_params = OmegaConf.to_container(self.cfg.rs.sampling_params, resolve=True)
+
+        self.to_offload_adam_states = self.cfg.rs.offload_adam_states
+        self.forward_micro_batch_size = self.cfg.rs.forward_micro_batch_size
+
+    def get_actor_forward_output_and_loss_func(self):
+        def fwd_output_and_loss_func(data_iterator, model):
+
+            batch = next(data_iterator)
+            response_tokens = batch["response_tokens"]
+            mask = batch["mask"]
+
+            attention_mask, _, position_ids = get_ltor_masks_and_position_ids(
+                data=response_tokens,
+                eod_token=self.tokenizer.eos_id,
+                reset_position_ids=False,
+                reset_attention_mask=False,
+                eod_mask_loss=False,
+            )
+
+            batch = {
+                "tokens": response_tokens,
+                "attention_mask": attention_mask,
+                "position_ids": position_ids,
+                "mask": mask,
+            }
+
+            required_keys = set()
+            if parallel_state.get_pipeline_model_parallel_world_size() == 1:
+                required_keys.update(batch.keys())
+            else:
+                required_keys.add("attention_mask")
+
+                if parallel_state.is_pipeline_first_stage():
+                    required_keys.update(("tokens", "position_ids"))
+
+                if parallel_state.is_pipeline_last_stage():
+                    required_keys.update(("tokens", "advantages", "mask", "prev_log_probs"))
+
+            batch = {key: val.cuda(non_blocking=True) if key in required_keys else None for key, val in batch.items()}
+
+            parallel_logits = model(batch["tokens"], batch["position_ids"], batch["attention_mask"], labels=None,)
+
+            def loss_func(parallel_logits):
```

Comment:
« O: » Just a note that this loss seems to suffer from the same issue as MegatronGPTModel where it depends on the mbs. This doesn’t need to be fixed here as we still need to fix it in NeMo first, but please add a TODO that points to the NeMo issue so as to keep a reference to it (search among the latest issues I created on NeMo)